### DISCUSSION TITLE
Add a threshold or delay so that Frigate does not consider an event ended before the tracked object(s) are actually obscured or out of frame.

### DISCUSSION THREAD

**Discussion by tyzbit** [blakeblackshear/frigate#8437] [General]
**Describe what you are trying to accomplish and why in non technical terms**
The issue I am seeing is objects are being detected, going below detection threshold and then being detected again which results in two events (immediately, I believe there is functionality that deduplicates events after the fact).

I want to be able to delay the end of an event despite the loss of tracked objects so that if an object is identified, then the score falls below `threshold`, then the object is detected again, Frigate considers that object to be a continuation of the previous event, not a new one.

**Describe the solution you'd like**
`threshold` could be renamed `initial_threshold` (suggested, not required - breaking changes suck) and a new config option `cutoff_threshold` could be added. An event would be created once the score was above `initial_threshold`, but while the object is detected its score is compared to `cutoff_threshold`. If the score falls below the `cutoff_threshold`, the event is considered ended.

This would allow for a high `initial_threshold` to prevent false positives (and perhaps delay initial detection by a small amount) but a lower `cutoff_threshold` so that subsequent detections that may fall below `initial_threshold` can be counted towards an event **continuing**, but not **starting**. The end result should hopefully be more contiguous events.

**Describe alternatives you've considered**
A camera setting, perhaps named `event_end_delay` where the value is a `float` that represents seconds for Frigate to wait before ending an event. If an object of the same type as the previous object that created the event is detected within `event_end_delay`, Frigate should continue tracking it as if it was part of the original event.

**Additional context**
None

**Comment by NickM-27** [comment]
I am going to move this to a discussion because there is a misunderstanding here on how frigate is working, and this feature request is not really applicable. Hopefully we can figure out the behavior you are seeing

An event is synonymous with a tracked object. The event must have a median score of > threshold for the event to be saved, but if the object's scores goes below threshold this will not create a new event. The only point at which an event is considered ended is when the object tracker hit_count goes to 0 because the object was not matched to a detection (because it was obscured, out of the frame, etc.)

**Reply by tyzbit** [reply]
That's consistent with my understanding. My desire is for Frigate not to create a new event every time an object's score drops below the threshold for however many frames and then back above it, but to hold onto the event in case the object is tracked again shortly thereafter.

**Reply by Mincka** [reply]
You mean, even if the object is completely out of the frame and you would accept the risk of having two different persons or cars to be merged in the same event, that's right?

I think this idea of "loose" event detection could be helpful and also applicable for other "complex" events with a group of persons. They generate multiple events for each person and multiple times for the same ones because they are hard to track. In the end, we just want to capture the timeframe when people were present, not overly excessive event handling.

In your case, your would like that the events would be "merged" if the same type of object reappears during the `max_disappeared` period.

**Reply by NickM-27** [reply]
@tyzbit again, what you're describing is not how frigate works. When an object score goes below threshold and above again it does not create a new event. It is the same event as before until the object is lost. The only way it would behave this way is if your min_score is too high or the same as threshold, which is not recommended at all

**Reply by NickM-27** [reply]
@Mincka I've not seen this behavior on the latest commits, I walked next to my wife from far away on the camera and it tracked us both on 2 events the entire way to the front door and also never ID swapped. 

The tracker is also much smarter than before and in my experience cars that leave and come at the same part of the image are not confused. Also, there are more features like reID that we have not started using yet that can make this even smarter.

**Reply by Mincka** [reply]
I can confirm that tracking has been immensely improved in latest Frigate version.

**Reply by tyzbit** [reply]
@NickM-27 I understand that's not how Frigate is designed to work but what actually happens on my cameras is Frigate correctly detects an object and creates an event, the object becomes obscured (or rotates, goes in  **and the object is lost (even though it is often still visible, at least partially)** and Frigate ends the event, then the object is detected again and a new event is created.

I have run with default thresholds and min_score and I have customized them but unfortunately it is not possible to filter out false positives while at the same time allowing for an object that passes the initial threshold to be only one event because the score will fall below the threshold. I know the models can and will be improved but this will continue to happen until they are essentially flawless.

**Reply by NickM-27** [reply]
> allowing for an object that passes the initial threshold to be only one event because the score will fall below the threshold

You are still misunderstanding how this is working. A new event being created has nothing to do with the object score falling below threshold. 

If the object is detected and an event is created but then that object becomes obscured enough that it is not detected for enough frames that its hit_count goes to 0 then the event must be ended. You will know this is the case watching the debug live view because when the object is detected it will have a standard color box that is "medium" thickness. When the object is not detected it will have a dark blue box that is thin. 

Every frame that the object is not detected its hit count goes down by one, every frame the object is detected its hit count goes up by 1. If the hit count makes it to 0 then the object is considered lost and the event ends. If the object is later detected it will create a new event. There is no way (currently) for frigate to know this is the same object as before so it must create a new event. 

But this is what I was referring to with the reID feature. In the future the tracker will be able to get the physical characteristics of the object so even when an object is obscured and lost, once it comes back into the frame (within a configured amount of time) it will be able to be matched with the existing event due to matching physical characteristics.

**Reply by NickM-27** [reply]
Seeing a screen recording of the debug live view would make it easier to see exactly what you are seeing

**Reply by tyzbit** [reply]
Thank you for clarifying, I was trying to describe the behavior I was seeing without knowing about the hit count metric. I think the features you're describing with the physical characteristics paired with a configurable amount of time for an object to return will satisfy my use case.

**Reply by joncar** [reply]
I've been seeing the same thing as tyzbit on the 0.13-beta3 and had been considering implementing this exact same thing with the same misunderstanding. To make it concrete an example of what I've been seeing lately (but wasn't seeing on 0.12): yesterday I had 11 car events in a row for a never-occluded stationary car. Most of the events are in alternating pattern of 9x% then 85% then 9x% then 85%. Regardless of the design of the fix, I'd like to find a way to fix this that doesn't involve lowering the initial threshold something much reach before being declared an object.

It might be that all of this will go away when I train my next Frigate+ model as I've since fed it a lot of examples of this car stationary at this location on the driveway which might help with what I assume are too many low %/entirely missed frames in a row.

FWIW, here's my relevant config snippets:
```
objects:
  filters:
    car:
      min_score: .65
      threshold: .80
detect:
  max_disappeared: 100
```

Reading your comments @NickM-27 I wonder if I should *lower* min_score more in the hopes of getting more object detected frames.

**Reply by NickM-27** [reply]
@joncar I would suggest using the recordings timeline on the events page, I would bet this has nothing to do with the object score and is probably due to a car driving behind your stationary car and the tracker is getting confused which is which.

**Reply by NickM-27** [reply]
and again, seeing the debug live view would be really helpful to understand what is going on. I have a car parked in my driveway and the event lasted the entire 18 hours it was there until it left again.

**Reply by joncar** [reply]
That's a good thought. But I live in a pretty rural place where cars rarely drive in the camera's background. I reviewed a bunch of the recordings and they are pretty much just the car sitting there with nothing changing around it. And the timeline ends with (naturally) that the object is "leaving" despite no change in the recorded image.

I didn't have much time to watch the debug view live today, but I did for a few minutes. A trend I noticed is it would detect the car with high confidence consistently after frigate startup. And then it would suddenly decay down to 50% and lose the object. I saw something similar when motion (nowhere near the car) caused the detector to start back up. It was almost like when it transitioned from motion detect to stationary detection, that transition was associated with a big drop in the object score.

I'm going to try to watch it more live tonight.

**Reply by NickM-27** [reply]
I'd suggest enabling regions, that will give important info

**Reply by kirsch33** [reply]
For what it’s worth I have experienced what @joncar described as well with cars in my driveway and not being occluded at any point. Just randomly being lost after a few hours for no real reason. I’d be curious to figure out how what’s going on to cause this. 

I did notice a huge improvement in maintaining my parked cars between 0.13 Beta 2 to 3 but still not quite there on my setup for tracking it the entire time it’s present.

**Reply by NickM-27** [reply]
we'll need to see the debug view particularly with regions enabled. Seeing the moment of when it stops being detected will provide the info on what happened, like I said I've personally not had any issues after the latest changes

![Screen Shot 2023-11-04 at 06 56 51 AM](https://github.com/blakeblackshear/frigate/assets/14866235/993013fe-9e84-4132-9321-9aa5a8cc9345)

**Reply by joncar** [reply]
Now that I brought it up here, it held the lock on the car for the 5 hours I parked it in the driveway today. But I think I understand why. Last night as I was debugging it was having trouble detecting in the dark at all so I 1) bumped up the detect stream's resolution a fair bit and 2) made sure the width and height in the frigate config exactly matched the stream's (it was previously slightly off from what the stream was). Now it is scoring the car at 99% consistently, so naturally all object disappearing problems went away.
